src/sentry/issues/occurrence_consumer.py

```diff
@@ -356,3 +362,56 @@ def _process_message(
             txn.set_tag("result", "error")
             raise InvalidEventPayloadError(e)
     return None
+
+
+def _process_batch(worker: ThreadPoolExecutor, message: Message[ValuesBatch[KafkaPayload]]):
+    """
+    Receives batches of occurrences. This function will take the batch
+    and group them together by fingerprint (ensuring order is preserved) and
+    execute each group using a ThreadPoolWorker.
+
+    By batching we're able to process occurrences in parallel while guaranteeing
+    that no occurrences are processed out of order per group.
+    """
+    batch = message.payload
+
+    occcurrence_mapping: Mapping[str, list[Mapping[str, Any]]] = defaultdict(list)
+
+    for item in batch:
+        assert isinstance(item, BrokerValue)
+
+        try:
+            payload = json.loads(item.payload.value, use_rapid_json=True)
+        except Exception:
+            logger.exception("Failed to unpack message payload")
+            continue
+        # group by the fingerprint, there should only be one of them
+        partition_key: str = payload["fingerprint"][0] if payload["fingerprint"] else ""
+        occcurrence_mapping[partition_key].append(payload)
+
+    # Number of occurrences that are being processed in this batch
+    metrics.gauge("occurrence_consumer.checkin.parallel_batch_count", len(batch))
+
+    # Number of groups we've collected to be processed in parallel
+    metrics.gauge("occurrence_consumer.checkin.parallel_batch_groups", len(occcurrence_mapping))
+    # Submit occurrences for processing
+    with sentry_sdk.start_transaction(op="process_batch", name="occurrence.occurrence_consumer"):
+        futures = [
+            worker.submit(process_occurrence_group, group) for group in occcurrence_mapping.values()
+        ]
+        wait(futures)
+
+
+def process_occurrence_group(items: list[Mapping[str, Any]]):
+    """
+    Process a group of related occurrences (all part of the same group)
+    completely serially.
+    """
+    for item in items:
+        cache_key = f"occurrence_consumer.process_occurrence_group.{item['event_id']}"
+        if cache.get(cache_key):
+            logger.info("Skipping processing of occurrence %s due to cache hit", item["event_id"])
+            continue
+        _process_message(item)
+        # just need a 300 second cache
+        cache.set(cache_key, 1, 300)
```

Comment:
This will process both `occurrences` and `status_change` right? Status change doesn't have an event_id https://github.com/getsentry/sentry/blob/eb55184602940adb1fc09a9c0e33515cac92c7d8/src/sentry/issues/status_change_message.py#L8-L12 Maybe we should add an id here?

For the occurrence, I'd recommend using the `id` over the `event_id` https://github.com/getsentry/sentry/blob/bf7faa720d72bf8484bef9e9f65d1bd4888d2c7d/src/sentry/issues/issue_occurrence.py#L26-L41

The `event_id` might be shared across multiple occurrences, for example if multiple performance issues were detected in a single transaction.